Userland/Libraries/LibGfx/ImageFormats/WebPLoader.cpp

```diff
@@ -1,624 +1,173 @@
+    ScopeGuard guard { [=]() { WebPMuxDelete(mux); } };
+
+    uint32_t flag = 0;
+    WebPMuxError err = WebPMuxGetFeatures(mux, &flag);
+    if (err != WEBP_MUX_OK)
+        return Error::from_string_literal("Failed to get webp features");
+
+    if (flag & ICCP_FLAG) {
+        WebPData icc_profile;
+        err = WebPMuxGetChunk(mux, "ICCP", &icc_profile);
+        if (err != WEBP_MUX_OK)
+            return Error::from_string_literal("Failed to get ICCP chunk of webp");
+
+        context.icc_data.resize(icc_profile.size);
+        memcpy(context.icc_data.data(), icc_profile.bytes, icc_profile.size);
+    }
+
+    context.state = WebPLoadingContext::State::HeaderDecoded;
     return {};
 }
 
-static ErrorOr<ImageData> decode_webp_animation_frame_image_data(ANMFChunk const& frame)
+static ErrorOr<void> decode_webp_image(WebPLoadingContext& context)
 {
-    ReadonlyBytes chunks = frame.frame_data;
-    auto chunk = TRY(RIFF::Chunk::decode_and_advance(chunks));
-
-    Optional<RIFF::Chunk> alpha, image_data;
-    if (chunk.id() == "ALPH"sv) {
-        alpha = chunk;
-        chunk = TRY(RIFF::Chunk::decode_and_advance(chunks));
-    }
-    if (chunk.id() == "VP8 "sv || chunk.id() == "VP8L"sv)
-        image_data = chunk;
-
-    return decode_webp_set_image_data(move(alpha), move(image_data));
-}
-
-static ErrorOr<NonnullRefPtr<Bitmap>> decode_webp_image_data(WebPLoadingContext& context, ImageData const& image_data)
-{
-    if (image_data.image_data_chunk.id() == "VP8L"sv) {
-        VERIFY(!image_data.alpha_chunk.has_value());
-        auto vp8l_header = TRY(decode_webp_chunk_VP8L_header(image_data.image_data_chunk.data()));
-
-        // Check that the VP8X header alpha flag matches the VP8L header alpha flag.
-        // FIXME: For animated images, if VP8X has alpha then at least one frame should have alpha. But we currently don't check this for animations.
-        if (context.first_chunk->id() == "VP8X" && !context.animation_frame_chunks_data.has_value() && context.vp8x_header.has_alpha != vp8l_header.is_alpha_used)
-            return Error::from_string_literal("WebPImageDecoderPlugin: VP8X header alpha flag doesn't match VP8L header");
-
-        return decode_webp_chunk_VP8L_contents(vp8l_header);
-    }
-
-    VERIFY(image_data.image_data_chunk.id() == "VP8 "sv);
-    auto vp8_header = TRY(decode_webp_chunk_VP8_header(image_data.image_data_chunk.data()));
-    auto bitmap = TRY(decode_webp_chunk_VP8_contents(vp8_header, image_data.alpha_chunk.has_value()));
+    VERIFY(context.state >= WebPLoadingContext::State::HeaderDecoded);
 
-    if (image_data.alpha_chunk.has_value())
-        TRY(decode_webp_chunk_ALPH(image_data.alpha_chunk.value(), *bitmap));
+    if (context.has_animation) {
+        WebPAnimDecoderOptions anim_decoder_options {};
+        WebPAnimDecoderOptionsInit(&anim_decoder_options);
+        anim_decoder_options.color_mode = MODE_BGRA;
+        anim_decoder_options.use_threads = 1;
```

Comment:
Why do we set that rather than letting the default value?

Reply:
Default `color_mode` is RGBA, which is incompatible with `Gfx::Bitmap`. Use threads was just a decision in favor of performance, but maybe it would be better to be conservative until we have the plumbing in place throughout the code base to specify this?